common/src/main/java/com/wynntils/features/user/ChatMentionFeature.java

```diff
@@ -0,0 +1,129 @@
+ */
+package com.wynntils.features.user;
+
+import com.wynntils.core.config.Config;
+import com.wynntils.core.config.ConfigHolder;
+import com.wynntils.core.features.UserFeature;
+import com.wynntils.handlers.chat.event.ChatMessageReceivedEvent;
+import com.wynntils.utils.mc.ComponentUtils;
+import com.wynntils.utils.mc.McUtils;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.chat.MutableComponent;
+import net.minecraft.sounds.SoundEvents;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+
+public class ChatMentionFeature extends UserFeature {
+    @Config
+    public boolean markMention = true;
+
+    @Config
+    public boolean dingMention = true;
+
+    @Config
+    public String aliases = "";
+
+    private Pattern pattern = Pattern.compile(
+            "(?<!\\[)\\b(" + McUtils.mc().getUser().getName()
+                    + (aliases.length() > 0 ? "|" + aliases.replace(",", "|") : "") + ")\\b(?!:|])",
+            Pattern.CASE_INSENSITIVE);
+
+    @Override
+    protected void onConfigUpdate(ConfigHolder configHolder) {
+        pattern = Pattern.compile(
+                "(?<!\\[)\\b(" + McUtils.mc().getUser().getName()
+                        + (aliases.length() > 0 ? "|" + aliases.replace(",", "|") : "") + ")\\b(?!:|])",
+                Pattern.CASE_INSENSITIVE);
+    }
+
+    @SubscribeEvent()
+    public void onChat(ChatMessageReceivedEvent e) {
+        Component message = e.getMessage();
+
+        Matcher looseMatcher = pattern.matcher(ComponentUtils.getUnformatted(message));
+
+        if (looseMatcher.find()) {
+            if (markMention) {
+                e.setMessage(rewriteComponentWithHighlight(message));
+            }
+            if (dingMention) {
+                McUtils.playSound(SoundEvents.NOTE_BLOCK_PLING.value());
+            }
+        }
+    }
+
+    private Component rewriteComponentWithHighlight(Component comp) {
+        MutableComponent curr = MutableComponent.create(comp.getContents()).withStyle(comp.getStyle());
+        // .getString() is used here as it gives formattingchars when those exist. It is needed for guild messages
+        // because wynn still uses legacy coloring for it.
+        String text = curr.getString();
+
+        if (text == "") {
+            // case: component has no text -> pass the component thru and process its siblings
+            List<Component> sib = comp.getSiblings();
+            if (sib.size() == 0) {
+                // no siblings -> this is an end component return self
+                return comp;
+            } else {
+                // some siblings -> check them
+                for (Component c : sib) {
+                    curr.append(rewriteComponentWithHighlight(c));
+                }
+            }
```

Comment:
>  clearly split it into two halves, the "have no text" and "have text" parts.

To do that, move the `return curr` here, and drop the `else {`.

Also, I'm thinking maybe you could even drop the `if (sib.size() == 0)` check and just have a for loop and a `return curr`. If I read the code correctly, `curr` is at this point essentially the same as `comp`.